zqd/ingest/archivepcap.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/brimsec/zq/pkg/ctxio"
+	"github.com/brimsec/zq/pkg/iosrc"
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio"
+	"github.com/brimsec/zq/zio/ndjsonio"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zqd/api"
+	"github.com/brimsec/zq/zqd/pcapanalyzer"
+	"github.com/brimsec/zq/zqd/pcapstorage"
+	"github.com/brimsec/zq/zqd/storage"
+	"golang.org/x/sync/errgroup"
+)
+
+type archivePcapOp struct {
+	cleanupfns     []func()
+	err            error
+	done           chan struct{}
+	pcapuri        iosrc.URI
+	pcapstore      *pcapstorage.Store
+	store          storage.Storage
+	zeek, suricata pcapanalyzer.Launcher
+	zctx           *resolver.Context
+
+	// snap not used for archive store pcap ingest. Here for functional parity
+	// with legacyPcapOp. Can be removed once filestore has been deprecated
+	snap chan struct{}
+
+	// stat fields
+	startTime      nano.Ts
+	pcapBytesTotal int64
+	pcapBytesRead  int64
+	recordCounter  *recordCounter
+}
+
+func newArchivePcapOp(ctx context.Context, logstore storage.Storage, pcapstore *pcapstorage.Store, pcapuri iosrc.URI, suricata, zeek pcapanalyzer.Launcher) (PcapOp, []string, error) {
+	info, err := iosrc.Stat(ctx, pcapuri)
+	if err != nil {
+		return nil, nil, err
+	}
+	warn := make(chan string)
+	go func() {
+		err = pcapstore.Update(ctx, pcapuri, warn)
+		close(warn)
+	}()
+	var warnings []string
+	for w := range warn {
+		warnings = append(warnings, w)
+	}
+	if err != nil {
+		return nil, warnings, err
+	}
+	p := &archivePcapOp{
+		startTime:      nano.Now(),
+		pcapBytesTotal: info.Size(),
+		pcapstore:      pcapstore,
+		store:          logstore,
+		pcapuri:        pcapuri,
+		recordCounter:  &recordCounter{},
+		done:           make(chan struct{}),
+		snap:           make(chan struct{}),
+		suricata:       suricata,
+		zeek:           zeek,
+		zctx:           resolver.NewContext(),
+	}
+	go func() {
+		p.err = p.run(ctx)
+		for _, fn := range p.cleanupfns {
+			fn()
+		}
+		close(p.done)
+	}()
+	return p, warnings, nil
+}
+
+func (p *archivePcapOp) run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pcapfile, err := iosrc.NewReader(ctx, p.pcapuri)
+	if err != nil {
+		return err
+	}
+	defer pcapfile.Close()
+	// Keeps track of bytes read from pcapfile.
+	r := io.TeeReader(pcapfile, p)
+
+	group, zreaders, err := p.runAnalyzers(ctx, r)
+	if err != nil {
+		return err
+	}
+	combiner := zbuf.NewCombiner(zreaders, zbuf.RecordCompare(p.store.NativeDirection()))
+	defer combiner.Close()
+	// track stats on records produced from analyzers.
+	p.recordCounter.reader = combiner
+
+	if err := p.store.Write(ctx, p.zctx, p.recordCounter); err != nil {
+		return err
+	}
+	return group.Wait()
+}
+
+func (p *archivePcapOp) runAnalyzers(ctx context.Context, pcapstream io.Reader) (*errgroup.Group, []zbuf.Reader, error) {
+	group, ctx := errgroup.WithContext(ctx)
```

Comment:
Consider moving the group creation up to `run()`, and passing it here, so that `run()` creates the cancellation context & the group that encompass the lifecycle of the operation.

Reply:
Sounds good.